File: ios/Sources/RCCore/State/Timeline.swift
import Foundation

/// Amendment A12: a message this app has sent that the device has not echoed
/// back yet. The app mints the `session.send` request id and the device returns
/// it as the `user_message` block id, so the row can be shown at once and
/// replaced in place when the event arrives.
public struct OptimisticMessage: Identifiable, Sendable, Equatable {
    /// The `session.send` request id, which becomes the device's `block_id`.
    public let id: String
    public let text: String
    public let attachments: [AttachmentInfo]
    /// When the send left the app, for the unconfirmed timeout.
    public let sentAt: Date

    public init(id: String, text: String, attachments: [AttachmentInfo] = [], sentAt: Date = Date()) {
        self.id = id
        self.text = text
        self.attachments = attachments
        self.sentAt = sentAt
    }

    /// How long a send may wait before the row stops claiming it is on its way.
    /// Well past any request timeout: what has not been confirmed by now is not
    /// slow, it is lost.
    public static let unconfirmedAfter: TimeInterval = 60

    public func isUnconfirmed(at now: Date = Date()) -> Bool {
        now.timeIntervalSince(sentAt) >= Self.unconfirmedAfter
    }

    /// Seconds left before this row says so, or zero once it has.
    public func remainingBeforeUnconfirmed(at now: Date = Date()) -> TimeInterval {
        max(0, Self.unconfirmedAfter - now.timeIntervalSince(sentAt))
    }
}

/// One renderable row. Block events collapse into a single entry keyed by
/// `block_id`; everything else gets a synthetic id derived from its `seq`.
public struct TimelineEntry: Identifiable, Sendable, Equatable {
    public let id: String
    /// Position in the transcript. Amendment A8: a block that keeps streaming
    /// takes the seq of its first appearance, so it holds its place while its
    /// `seq` rises.
    public private(set) var seq: Int
    /// The raw seq of the event that created this entry, which is what a
    /// history cursor may be built from.
    public private(set) var eventSeq: Int
    /// The newest seq applied to this entry, for late-frame rejection.
    public private(set) var latestSeq: Int
    public private(set) var ts: Int64
    public let parentID: String?
    public private(set) var body: SessionEventBody
    /// Accumulated streaming text for `assistant_text` and `thinking`.
    public private(set) var text: String
    /// Amendment A12: set on a row this app has sent and the device has not
    /// confirmed. Such a row holds no device `seq` and is never stored in the
    /// transcript, so nothing it does can move the replay cursor.
    public private(set) var pending: OptimisticMessage?

    init(event: SessionEvent, id: String) {
        self.id = id
        seq = event.orderSeq
        eventSeq = event.seq
        latestSeq = event.seq
        ts = event.ts
        parentID = event.parentBlockID
        body = event.body
        text = Self.initialText(event.body)
    }

    /// The row a sent message takes until the device's own event replaces it.
    init(pending: OptimisticMessage) {
        id = pending.id
        // Nothing the device sent can sort after this, and nothing here can be
        // mistaken for a real event seq.
        seq = Int.max
        eventSeq = Int.max
        latestSeq = Int.max
        ts = Int64(pending.sentAt.timeIntervalSince1970 * 1000)
        parentID = nil
        body = .userMessage(UserMessagePayload(text: pending.text,
                                               attachments: pending.attachments,
                                               source: .remote))
        text = ""
        self.pending = pending
    }

    private static func initialText(_ body: SessionEventBody) -> String {
        switch body {
        case .assistantText(let payload), .thinking(let payload):
            payload.text ?? payload.delta ?? ""
        default:
            ""
        }
    }

    /// Apply a later event for the same block. A `delta` appends; anything else
    /// replaces the entry wholesale, per the protocol's replacement rule.
    mutating func merge(_ event: SessionEvent) {
        latestSeq = event.seq
        // A later event may be the first to name the block's origin, and an
        // earlier one may arrive out of order; the earliest position wins.
        seq = min(seq, event.orderSeq)
        eventSeq = min(eventSeq, event.seq)
        ts = event.ts
        switch event.body {
        case .assistantText(let payload), .thinking(let payload):
            if let delta = payload.delta, payload.text == nil {
                text += delta
            } else if let full = payload.text {
                text = full
            }
        default:
            text = Self.initialText(event.body)
        }
        body = event.body
    }

    public var userMessage: UserMessagePayload? {
        if case .userMessage(let payload) = body { payload } else { nil }
    }
    public var assistantText: StreamTextPayload? {
        if case .assistantText(let payload) = body { payload } else { nil }
    }
    public var thinking: StreamTextPayload? {
        if case .thinking(let payload) = body { payload } else { nil }
    }
    public var toolCall: ToolCallPayload? {
        if case .toolCall(let payload) = body { payload } else { nil }
    }
    public var approval: ApprovalPayload? {
        if case .approval(let payload) = body { payload } else { nil }
    }
    public var question: QuestionPayload? {
        if case .question(let payload) = body { payload } else { nil }
    }
    public var notice: NoticePayload? {
        if case .notice(let payload) = body { payload } else { nil }
    }
    public var errorPayload: ErrorPayload? {
        if case .error(let payload) = body { payload } else { nil }
    }
    public var turnStarted: TurnStartedPayload? {
        if case .turnStarted(let payload) = body { payload } else { nil }
    }
    public var turnCompleted: TurnCompletedPayload? {
        if case .turnCompleted(let payload) = body { payload } else { nil }
    }

    /// Rows a sub-agent produced hang under their parent tool call.
    public var isNested: Bool { parentID != nil }

    /// Entries that carry no visible content of their own.
    public var isRenderable: Bool {
        switch body {
        case .todos, .status, .meta, .queue: false
        case .assistantText, .thinking: !text.isEmpty || !isStreaming
        default: true
        }
    }

    public var isStreaming: Bool {
        switch body {
        case .assistantText(let payload), .thinking(let payload): !payload.done
        case .toolCall(let payload): payload.status == .running
        default: false
        }
    }
}

/// The ordered transcript of one session, plus the snapshots that ride along
/// with it (todos, queue). Pure value type: the store owns one and the
/// verification executable drives it directly.
public struct Timeline: Sendable, Equatable {
    public private(set) var entries: [TimelineEntry] = []
    /// Highest seq applied from the live stream. Late or duplicate frames at or
    /// below this are dropped.
    public private(set) var lastSeq = 0
    public private(set) var todos: [TodoItem] = []
    public private(set) var queue: [QueuedMessage] = []
    public private(set) var hasMoreHistory = true
    public private(set) var historyLoaded = false
    /// True when a `seq` was skipped, so the transcript is missing an event.
    /// The gateway drops frames over 64 KiB and a stream buffer can overrun, so
    /// a hole is part of the design and has to be noticed and repaired.
    public private(set) var hasGap = false
    /// Amendment A12: sent, not yet echoed by the device. Kept apart from
    /// `entries` because these rows carry no `seq`: they must not move the
    /// replay cursor, must not be a history boundary, and always sort last.
    public private(set) var optimistic: [OptimisticMessage] = []

    private var index: [String: Int] = [:]
    /// Children keyed by parent block, so a tool row does not scan the whole
    /// transcript on every delta.
    private var childIndex: [String: [String]] = [:]
    /// The seq each snapshot came from. A snapshot found while paging history
    /// is older than the live one and must not overwrite it (amendment A6).
    private var todosSeq = 0
    private var queueSeq = 0

    public init() {}

    /// The cursor for `session.history`: the oldest real event seq held, not a
    /// block's position, so paging cannot skip an event.
    public var oldestSeq: Int? { entries.map(\.eventSeq).min() }

    public var renderable: [TimelineEntry] { entries.filter(\.isRenderable) }

    /// Top-level rows, then the sends the device has not confirmed yet. A
    /// pending row carries the `user_message` it is about to become, so the
    /// transcript renders it like any other (amendment A12).
    public var roots: [TimelineEntry] {
        var rows = entries.filter { $0.isRenderable && !$0.isNested }
        rows.append(contentsOf: optimistic.filter { index[$0.id] == nil }.map(TimelineEntry.init(pending:)))
        return rows
    }

    public func children(of blockID: String) -> [TimelineEntry] {
        (childIndex[blockID] ?? []).compactMap(entry(id:)).filter(\.isRenderable)
    }

    public func entry(id: String) -> TimelineEntry? {
        index[id].map { entries[$0] }
    }

    /// The newest approval or question still waiting on the user.
    public var pendingRequest: TimelineEntry? {
        entries.last { entry in
            entry.approval?.status.isActionable == true || entry.question?.status.isActionable == true
        }
    }

    /// Apply one live event. Returns false when the event was dropped as stale.
    ///
    /// A seq that skips ahead means an event never arrived. The entry is still
    /// applied, but the timeline is marked so the store can refill from the
    /// gateway rather than render a transcript that is quietly missing a step.
    @discardableResult
    public mutating func apply(_ event: SessionEvent) -> Bool {
        guard event.seq > lastSeq else { return false }
        if lastSeq > 0, event.seq > lastSeq + 1 { hasGap = true }
        lastSeq = event.seq
        reconcileOptimistic(with: event)
        absorb(event)
        return true
    }

    /// Called once the store has refilled from `session.subscribe` or history.
    public mutating func clearGap() { hasGap = false }

    /// Adopt a cursor recovered from the offline cache, so a warm open can
    /// subscribe with `since_seq` instead of throwing the transcript away.
    public mutating func adoptCursor(_ seq: Int) {
        guard seq > lastSeq else { return }
        lastSeq = seq
    }

    /// Merge one page of older history. History never advances `lastSeq`, and
    /// an entry already held from the live stream wins over an older copy.
    public mutating func prependHistory(_ events: [SessionEvent], hasMore: Bool) {
        for event in events.sorted(by: { $0.seq < $1.seq }) {
            reconcileOptimistic(with: event)
            let key = Self.key(for: event)
            if let position = index[key] {
                guard event.seq > entries[position].latestSeq else { continue }
                entries[position].merge(event)
            } else {
                index[key] = entries.count
                entries.append(TimelineEntry(event: event, id: key))
            }
            absorbSnapshot(event)
        }
        sortEntries()
        hasMoreHistory = hasMore
        historyLoaded = true
    }

    /// Replace the whole transcript, used when the gateway reports `resync`.
    ///
    /// The unconfirmed sends survive it: a message the user just typed must not
    /// vanish because the socket came back and the gateway asked for a reload.
    /// They are reconciled again as the reloaded events come through.
    public mutating func reset() {
        entries.removeAll()
        index.removeAll()
        childIndex.removeAll()
        hasGap = false
        todos.removeAll()
        queue.removeAll()
        lastSeq = 0
        todosSeq = 0
        queueSeq = 0
        hasMoreHistory = true
        historyLoaded = false
    }

    /// The queue snapshot from a `session.subscribe` reply, which has no seq of
    /// its own: it describes the session as of the current cursor.
    public mutating func applySubscribedQueue(_ pending: [QueuedMessage]) {
        guard lastSeq >= queueSeq else { return }
        queue = pending
        queueSeq = lastSeq
        dropQueuedOptimistic()
    }

    // MARK: - Amendment A12: sends the device has not confirmed

    /// Show a message the moment it is sent. Idempotent, so a Retry under the
    /// same request id reuses the row rather than adding a second one.
    public mutating func addOptimistic(_ message: OptimisticMessage) {
        guard index[message.id] == nil, !optimistic.contains(where: { $0.id == message.id }) else { return }
        optimistic.append(message)
    }

    /// Take a pending row away: the send was refused, or the queue owns it now.
    public mutating func removeOptimistic(_ id: String) {
        optimistic.removeAll { $0.id == id }
    }

    /// The pending rows that have waited too long to still claim they are on
    /// their way. Pure; the caller supplies the clock.
    public func unconfirmedOptimistic(at now: Date = Date()) -> [OptimisticMessage] {
        optimistic.filter { $0.isUnconfirmed(at: now) }
    }

    /// A queued message is represented by the queue row above the composer
    /// until the device dequeues it and emits the `user_message` under the same
    /// id, so showing both would show it twice.
    private mutating func dropQueuedOptimistic() {
        guard !optimistic.isEmpty, !queue.isEmpty else { return }
        let queued = Set(queue.map(\.id))
        optimistic.removeAll { queued.contains($0.id) }
    }

    /// Retire the pending row an incoming event confirms. Matching on the block
    /// id is amendment A12; matching one row by text is the fallback for a
    /// device that still mints its own ids, and never retires more than one.
    private mutating func reconcileOptimistic(with event: SessionEvent) {
        guard !optimistic.isEmpty else { return }
        let key = Self.key(for: event)
        if optimistic.contains(where: { $0.id == key }) {
            optimistic.removeAll { $0.id == key }
            return
        }
        guard case .userMessage(let payload) = event.body, payload.source == .remote,
              let position = optimistic.firstIndex(where: { $0.text == payload.text }) else { return }
        optimistic.remove(at: position)
    }

    /// Fold in the untruncated version of one block from `session.block`.
    public mutating func replaceBlock(with event: SessionEvent) {
        reconcileOptimistic(with: event)
        let key = Self.key(for: event)
        if let position = index[key] {
            entries[position].merge(event)
        } else {
            absorb(event)
        }
    }

    public mutating func markHistoryExhausted() {
        hasMoreHistory = false
        historyLoaded = true
    }

    private mutating func absorb(_ event: SessionEvent) {
        absorbSnapshot(event)
        switch event.body {
        case .todos, .status, .meta, .queue:
            return
        default:
            break
        }
        let key = Self.key(for: event)
        if let position = index[key] {
            let before = entries[position].seq
            entries[position].merge(event)
            if entries[position].seq != before { sortEntries() }
        } else {
            let entry = TimelineEntry(event: event, id: key)
            // The common case is an append at the end; only an out-of-order
            // position pays for a sort.
            let inOrder = entries.last.map { $0.seq <= entry.seq } ?? true
            index[key] = entries.count
            entries.append(entry)
            if let parent = event.parentBlockID { childIndex[parent, default: []].append(key) }
            if !inOrder { sortEntries() }
        }
    }

    private mutating func sortEntries() {
        entries.sort { ($0.seq, $0.eventSeq) < ($1.seq, $1.eventSeq) }
        reindex()
    }

    /// Snapshots replace the previous list, but only when they are at least as
    /// new as the one already held. Replayed events, history pages and live
    /// frames all pass through here, so the newest wins regardless of arrival
    /// order (amendment A6).
    private mutating func absorbSnapshot(_ event: SessionEvent) {
        switch event.body {
        case .todos(let payload):
            guard event.seq >= todosSeq else { return }
            todos = payload.items
            todosSeq = event.seq
        case .queue(let payload):
            guard event.seq >= queueSeq else { return }
            queue = payload.pending
            queueSeq = event.seq
            dropQueuedOptimistic()
        default:
            break
        }
    }

    private mutating func reindex() {
        index.removeAll(keepingCapacity: true)
        childIndex.removeAll(keepingCapacity: true)
        for (position, entry) in entries.enumerated() {
            index[entry.id] = position
            if let parent = entry.parentID { childIndex[parent, default: []].append(entry.id) }
        }
    }

    private static func key(for event: SessionEvent) -> String {
        event.blockID ?? "seq:\(event.seq)"
    }
}

File: web/src/stores/timeline.ts
/**
 * Pure timeline reducer: session events in, ordered renderable items out.
 *
 * Contract rules implemented here (PROTOCOL-FROZEN.md §4, §7, A8):
 *  - a later event with the same `block_id` fully replaces the earlier one,
 *    except `delta` events on `assistant_text` / `thinking`, which append;
 *  - events whose `seq` is not greater than the last applied one are dropped;
 *  - `status`, `meta`, `queue` and `todos` are session state, not timeline items;
 *  - events carrying `parent_block_id` belong under their parent tool row;
 *  - blocks are ordered by `first_seq ?? seq`, so a tool call that finishes long
 *    after it started keeps its place — live and after a reload.
 *
 * Amendment A12 adds one more: the app mints the `session.send` request id and
 * the device echoes it as the `user_message` block id, so the message is shown
 * the moment it is sent and the device's event replaces it under rule 1. Those
 * optimistic blocks are kept apart from `items` — they carry no `seq`, must not
 * move the replay cursor, and always sort after everything the device sent.
 */
import { eventKey } from '../lib/ids';
import type { Attachment, SessionEvent } from '../protocol/types';

export interface TimelineItem {
  key: string;
  /** Ordering key: `first_seq ?? seq` of the block's first appearance. */
  seq: number;
  ts: number;
  parentBlockId: string | null;
  event: SessionEvent;
  /** Set on a row the device has not confirmed yet. */
  pending?: OptimisticBlock;
}

/** A message this app has sent and the device has not echoed back yet (A12). */
export interface OptimisticBlock {
  /** The `session.send` request id, which becomes the device's `block_id`. */
  id: string;
  text: string;
  attachments: Attachment[];
  /** When the app sent it, for the unconfirmed timeout. */
  at: number;
  /**
   * What `session.send` answered, once it has. A `steered` message waits for the
   * agent to reach its next step (A14), which takes as long as it takes, so the
   * wait says nothing about delivery.
   */
  accepted?: 'sent' | 'steered';
}

export interface TimelineState {
  order: string[];
  items: Record<string, TimelineItem>;
  lastSeq: number;
  oldestSeq: number | null;
  /** Sent, not yet confirmed. Rendered after `order`, oldest first. */
  optimistic: OptimisticBlock[];
}

export function emptyTimeline(): TimelineState {
  return { order: [], items: {}, lastSeq: 0, oldestSeq: null, optimistic: [] };
}

/**
 * A fresh timeline for a resync that keeps the unconfirmed sends. Dropping them
 * would make a message the user just sent vanish on a reconnect.
 */
export function keepOptimistic(previous: TimelineState): TimelineState {
  return { ...emptyTimeline(), optimistic: previous.optimistic };
}

/** Event kinds that never produce a timeline item. */
const STATE_ONLY = new Set(['status', 'meta', 'queue', 'todos', 'turn_started']);

function isRenderable(event: SessionEvent): boolean {
  if (STATE_ONLY.has(event.kind)) return false;
  // A clean turn boundary needs no row; interrupted / failed turns do.
  if (event.kind === 'turn_completed') return event.stop_reason !== 'completed';
  return true;
}

function keyFor(event: SessionEvent): string {
  if ('block_id' in event && typeof event.block_id === 'string' && event.block_id) {
    return event.block_id;
  }
  return eventKey(event.kind, event.seq);
}

/**
 * Where the block belongs in the timeline. `first_seq` wins when the device
 * sends it; otherwise the earliest `seq` this app has seen for the block does.
 */
function orderSeq(event: SessionEvent, previous: TimelineItem | undefined): number {
  const declared = typeof event.first_seq === 'number' ? event.first_seq : event.seq;
  return previous ? Math.min(previous.seq, declared) : declared;
}

function streamingText(previous: SessionEvent | undefined, next: SessionEvent): string | undefined {
  if (next.kind !== 'assistant_text' && next.kind !== 'thinking') return undefined;
  if (next.delta === undefined) return next.text;
  const prevText =
    previous && (previous.kind === 'assistant_text' || previous.kind === 'thinking')
      ? (previous.text ?? '')
      : '';
  return prevText + next.delta;
}

/** Merge one event into an existing item (or create it). Returns null when not renderable. */
function mergeItem(previous: TimelineItem | undefined, event: SessionEvent): TimelineItem | null {
  if (!isRenderable(event)) return null;
  const key = keyFor(event);
  const text = streamingText(previous?.event, event);
  const merged: SessionEvent =
    text === undefined
      ? event
      : ({ ...event, text, delta: undefined } as SessionEvent);
  return {
    key,
    seq: orderSeq(event, previous),
    ts: event.ts,
    parentBlockId: event.parent_block_id ?? previous?.parentBlockId ?? null,
    event: merged,
  };
}

/**
 * Insert a key so `order` stays sorted by ordering seq. Scans from the end, so
 * the common case — a block that belongs last — costs one comparison.
 */
function insertOrdered(
  order: readonly string[],
  items: Record<string, TimelineItem>,
  key: string,
  seq: number,
): string[] {
  let index = order.length;
  while (index > 0) {
    const previous = items[order[index - 1] as string];
    if (previous !== undefined && previous.seq > seq) index -= 1;
    else break;
  }
  if (index === order.length) return [...order, key];
  const next = [...order];
  next.splice(index, 0, key);
  return next;
}

/* --------------------------------------------------------- optimistic (A12) */

/** How long an unconfirmed send waits before the row says so. */
export const UNCONFIRMED_AFTER_MS = 60_000;

/**
 * Whether a pending row has waited too long. Pure; the row needs only a clock.
 * §8 rule 7 reserves this for a send whose delivery is genuinely uncertain, so
 * a gateway answer settles it for good however long the agent then takes.
 */
export function isUnconfirmed(block: OptimisticBlock, now: number): boolean {
  if (block.accepted !== undefined) return false;
  return now - block.at >= UNCONFIRMED_AFTER_MS;
}

/** Show a message the moment it is sent. Idempotent, so Retry reuses the row. */
export function addOptimistic(state: TimelineState, block: OptimisticBlock): TimelineState {
  if (state.items[block.id] || state.optimistic.some((b) => b.id === block.id)) return state;
  return { ...state, optimistic: [...state.optimistic, block] };
}

/**
 * Record what `session.send` answered for a pending row. The row itself stays:
 * under A14 a `steered` message becomes a device block only when the agent
 * takes it, and a `sent` one when the device echoes it.
 */
export function markAccepted(
  state: TimelineState,
  id: string,
  accepted: 'sent' | 'steered',
): TimelineState {
  if (!state.optimistic.some((b) => b.id === id && b.accepted !== accepted)) return state;
  return {
    ...state,
    optimistic: state.optimistic.map((b) => (b.id === id ? { ...b, accepted } : b)),
  };
}

/** Drop a pending row: the send was refused, or its queue entry was removed. */
export function removeOptimistic(state: TimelineState, id: string): TimelineState {
  if (!state.optimistic.some((b) => b.id === id)) return state;
  return { ...state, optimistic: state.optimistic.filter((b) => b.id !== id) };
}

/**
 * Retire the pending rows a `queue` snapshot has taken over. A queued message
 * is represented by the row above the composer until the device dequeues it and
 * emits the `user_message` under the same id (A12), so showing both would show
 * it twice.
 */
export function dropQueued(state: TimelineState, ids: readonly string[]): TimelineState {
  if (state.optimistic.length === 0 || ids.length === 0) return state;
  const optimistic = state.optimistic.filter((block) => !ids.includes(block.id));
  return optimistic.length === state.optimistic.length ? state : { ...state, optimistic };
}

/**
 * Retire the pending rows an incoming event confirms. The id match is A12; the
 * text match is the fallback for a device that still mints its own block ids,
 * and only ever retires one row per event.
 */
function reconcile(optimistic: OptimisticBlock[], event: SessionEvent): OptimisticBlock[] {
  if (optimistic.length === 0) return optimistic;
  const key = keyFor(event);
  if (optimistic.some((b) => b.id === key)) return optimistic.filter((b) => b.id !== key);
  if (event.kind !== 'user_message' || event.source !== 'remote') return optimistic;
  const index = optimistic.findIndex((b) => b.text === event.text);
  if (index === -1) return optimistic;
  return optimistic.filter((_, i) => i !== index);
}

/** Apply a live event. Out-of-order or duplicate frames are dropped. */
export function applyEvent(state: TimelineState, event: SessionEvent): TimelineState {
  if (event.seq <= state.lastSeq) return state;
  const key = keyFor(event);
  const previous = state.items[key];
  const item = mergeItem(previous, event);
  const lastSeq = event.seq;
  const oldestSeq = state.oldestSeq === null ? event.seq : Math.min(state.oldestSeq, event.seq);
  const optimistic = reconcile(state.optimistic, event);
  if (!item) return { ...state, lastSeq, optimistic };

  const items = { ...state.items, [key]: item };
  let order = state.order;
  if (previous === undefined) {
    order = insertOrdered(order, items, key, item.seq);
  } else if (previous.seq !== item.seq) {
    // A replacement revealed an earlier `first_seq`: move the row back.
    order = insertOrdered(
      order.filter((k) => k !== key),
      items,
      key,
      item.seq,
    );
  }
  return { order, items, lastSeq, oldestSeq, optimistic };
}

export function applyEvents(state: TimelineState, events: readonly SessionEvent[]): TimelineState {
  return events.reduce(applyEvent, state);
}

/**
 * Merge an older history page (ascending `seq`, ending before the oldest known
 * event). Existing items win: they are at least as fresh as the history copy.
 */
export function mergeHistory(state: TimelineState, events: readonly SessionEvent[]): TimelineState {
  if (events.length === 0) return state;
  const items = { ...state.items };
  const added: TimelineItem[] = [];
  let oldest = state.oldestSeq;
  let optimistic = state.optimistic;
  for (const event of events) {
    if (oldest === null || event.seq < oldest) oldest = event.seq;
    // A page can carry the device's copy of a send this app is still showing.
    optimistic = reconcile(optimistic, event);
    const key = keyFor(event);
    if (items[key]) continue;
    const item = mergeItem(undefined, event);
    if (!item) continue;
    items[key] = item;
    added.push(item);
  }
  // Both sides are already sorted by ordering seq, so merge instead of sorting.
  added.sort((a, b) => a.seq - b.seq);
  const order = mergeSorted(added, state.order, items);
  return {
    order,
    items,
    lastSeq: Math.max(state.lastSeq, events[events.length - 1]?.seq ?? 0),
    oldestSeq: oldest,
    optimistic,
  };
}

function mergeSorted(
  added: readonly TimelineItem[],
  existing: readonly string[],
  items: Record<string, TimelineItem>,
): string[] {
  const out: string[] = [];
  let i = 0;
  let j = 0;
  while (i < added.length && j < existing.length) {
    const left = added[i] as TimelineItem;
    const right = items[existing[j] as string];
    if (right === undefined || left.seq <= right.seq) {
      out.push(left.key);
      i += 1;
    } else {
      out.push(existing[j] as string);
      j += 1;
    }
  }
  for (; i < added.length; i += 1) out.push((added[i] as TimelineItem).key);
  for (; j < existing.length; j += 1) out.push(existing[j] as string);
  return out;
}

/** Replace one block with its full, untruncated version (`session.block`). */
export function replaceBlock(state: TimelineState, event: SessionEvent): TimelineState {
  const key = keyFor(event);
  const previous = state.items[key];
  if (!previous) return state;
  return {
    ...state,
    items: { ...state.items, [key]: { ...previous, ts: event.ts, event } },
  };
}

export interface TimelineView {
  roots: TimelineItem[];
  children: Record<string, TimelineItem[]>;
}

/**
 * How much of the timeline is drawn (`docs/DESIGN.md` § "The timeline"). Simple
 * shows only what is written to the person; Detailed adds everything the agent
 * did on the way. The level is local to the app and never reaches the wire.
 */
export type TimelineDetail = 'simple' | 'detailed';

/** Kinds Simple never draws: not collapsed, not summarised, not counted. */
const WORKINGS = new Set(['thinking', 'tool_call']);

/** Kinds no level hides, wherever they sit: they are waiting for an answer. */
const ALWAYS_DRAWN = new Set(['approval', 'question']);

const drawnAt = (item: TimelineItem, detail: TimelineDetail): boolean =>
  detail === 'detailed' || !WORKINGS.has(item.event.kind);

/**
 * Split items into top-level rows and sub-agent children keyed by parent block,
 * then append the sends the device has not confirmed yet. A pending row carries
 * the `user_message` it will become, so the timeline renders it like any other.
 *
 * Simple drops the agent's workings along with everything nested under them,
 * except a card that needs an answer: it comes up to the top level rather than
 * going with the tool row that held it.
 */
export function selectView(state: TimelineState, detail: TimelineDetail): TimelineView {
  const roots: TimelineItem[] = [];
  const children: Record<string, TimelineItem[]> = {};
  for (const key of state.order) {
    const item = state.items[key];
    if (!item || !drawnAt(item, detail)) continue;
    const parent = item.parentBlockId ? state.items[item.parentBlockId] : undefined;
    if (parent && drawnAt(parent, detail)) {
      (children[parent.key] ??= []).push(item);
    } else if (!parent || ALWAYS_DRAWN.has(item.event.kind)) {
      roots.push(item);
    }
  }
  for (const block of state.optimistic) {
    if (state.items[block.id]) continue;
    roots.push(pendingItem(block));
  }
  return { roots, children };
}

function pendingItem(block: OptimisticBlock): TimelineItem {
  return {
    key: block.id,
    // Pending rows always sort last; they hold no device `seq`.
    seq: Number.MAX_SAFE_INTEGER,
    ts: block.at,
    parentBlockId: null,
    pending: block,
    event: {
      seq: 0,
      ts: block.at,
      kind: 'user_message',
      block_id: block.id,
      text: block.text,
      source: 'remote',
      ...(block.attachments.length > 0 ? { attachments: block.attachments } : {}),
    },
  };
}
